docs/source/VUDRM-key-provision.md

```diff
@@ -0,0 +1,442 @@
+
+encrypted_keys = "KEY_PROVIDER_RESPONSE"
+
+message, message_hash = encrypted_keys.split('|')
+
+pre_computed = client + message
+computed_hash = Digest::SHA1.new.hexdigest(shared_secret + pre_computed)
+
+if computed_hash == message_hash
+  decoded_message = Base64.strict_decode64(message)
+  decipher = OpenSSL::Cipher::Cipher.new('AES-128-ECB').decrypt
+  decipher.key = shared_secret[0..15]
+  unencrypted_keys = decipher.update(decoded_message) + decipher.final
+else
+  raise "Key provider response hash did not validate"
+end
+```
+
+##### GO
+
+```go
+package main
+
+import (
+	"crypto/aes"
+	"crypto/sha1"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"strings"
+)
+
+var (
+	Client = "CLIENT"
+	SharedSecret = "SHARED_SECRET"
+	KeyProviderResponse = "KEY_PROVIDER_RESPONSE"
+	PKCS5 = &pkcs5{}
+)
+
+type pkcs5 struct{}
+
+func main() {
+	s := strings.Split(KeyProviderResponse, "|")
+	encryptedMessage, hash := s[0], s[1]
+	generatedHash := generateKeyProviderHash(SharedSecret, Client, encryptedMessage)
+	if generatedHash == hash {
+		decryptedKeys := decrypt(encryptedMessage, SharedSecret)
+		fmt.Printf(decryptedKeys)
+	} else {
+		fmt.Printf("hash failed validation")
+	}
+}
+
+func generateKeyProviderHash(sharedSecret string, client string, encryptedMessage string) string {
+	v := fmt.Sprintf("%s%s%s", sharedSecret, client, encryptedMessage)
+	hasher := sha1.New()
+	hasher.Write([]byte(v))
+	return fmt.Sprintf("%x", hasher.Sum(nil))
+}
+
+func decrypt(encryptedData string, sharedSecret string) string {
+	source, _ := base64.StdEncoding.DecodeString(encryptedData)
+	key := []byte(sharedSecret)[0:16]
+
+	cipher, _ := aes.NewCipher([]byte(key))
+	buffer := make([]byte, len(source))
+	size := 16
+
+	for bs, be := 0, size; bs < len(source); bs, be = bs+size, be+size {
+		cipher.Decrypt(buffer[bs:be], source[bs:be])
+	}
+
+	plainBytes, _ := PKCS5.unPadding(buffer, 16)
+	return string(plainBytes)
+}
+
+func (p *pkcs5) unPadding(src []byte, blockSize int) ([]byte, error) {
+	srcLen := len(src)
+	paddingLen := int(src[srcLen-1])
+	if paddingLen >= srcLen || paddingLen > blockSize {
+		return nil, errors.New("padding size error")
+	}
+	return src[:srcLen-paddingLen], nil
+}
+```
+
+##### Python
+
+```python
+import base64
+import binascii
+import json
+import hashlib
+
+from Crypto.Cipher import AES
+
+client = "CLIENT"
+shared_secret = "SHARED_SECRET"
+key_provider_response = "KEY_PROVIDER_RESPONSE"
+
+print("Decrypting Key Provider Response...")
+split_key_provider_response = key_provider_response.split("|")
+pre_computed = (shared_secret + client + split_key_provider_response[0]).encode("utf-8")
+computed_hash = hashlib.sha1(pre_computed).hexdigest()
+if computed_hash == split_key_provider_response[1]:
+    key = shared_secret[0:16]
+    decoded_text = base64.b64decode(split_key_provider_response[0])
+    aes = AES.new(key.encode("utf8"), AES.MODE_ECB)
+    padded_text = aes.decrypt(decoded_text)
+    unpadded_text = padded_text[:-padded_text[-1]]
```

Comment:
Is this the same bug? Now I'm starting to wonder if this is a bug, maybe we always put 1 byte on the end of the message.

Reply:
It is padding. Might be a single byte. Depends on the difference between the length of the text to be encrypted and the next blocksize multiple. Also it works in production ;)